### PR TITLE
update the femnist example such that admm algorithms can be used

### PR DESCRIPTION
**Pull request by minseok-ryu**
The femnist example was written for algorithms belonging to the Federated() class. 
To use another classes such as IIADMM() and ICEADMM(), the femnist example should be revised.

### CODE REVIEW

**Comment by minseok-ryu** [comment]
@yim0331 ADMM() and ICEADMM() can be used for training CNN model, but the CUDA out-of-memory issue was encountered when I tried to use the pretrained resnet18. Did you have the similar issue before?

**Comment by minseok-ryu** [comment]
For more details, I used the following command to run:
```
mpiexec -np 4 --mca opal_cuda_support 1 python ./femnist.py --model resnet18 --pretrained 1 --fed="IIADMM()"
```
The error looks like the following:
```
----------Loaded Datasets and Model----------Elapsed Time= 22.210840225219727
----------Loaded Datasets and Model----------Elapsed Time= 22.211190700531006
----------Loaded Datasets and Model----------Elapsed Time= 22.21150827407837
----------Loaded Datasets and Model----------Elapsed Time= 22.21161651611328
./femnist.py:81: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:245.)
  torch.FloatTensor(test_data_input), torch.tensor(test_data_label)
./femnist.py:81: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:245.)
  torch.FloatTensor(test_data_input), torch.tensor(test_data_label)
Traceback (most recent call last):
  File "./femnist.py", line 169, in <module>
    main()
  File "./femnist.py", line 162, in main
    rm.run_client(cfg, comm, model, loss_fn, args.num_clients, train_datasets)
  File "/gpfs/fs1/home/mryu/APPFL/src/appfl/run_mpi.py", line 283, in run_client
    ls = client.update()                            
  File "/gpfs/fs1/home/mryu/APPFL/src/appfl/algorithm/iiadmm.py", line 160, in update
    loss.backward()
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/_tensor.py", line 487, in backward
    torch.autograd.backward(
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/autograd/__init__.py", line 200, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 2.00 MiB (GPU 0; 39.59 GiB total capacity; 11.76 GiB already allocated; 4.19 MiB free; 11.94 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
Traceback (most recent call last):
  File "./femnist.py", line 169, in <module>
    main()
  File "./femnist.py", line 162, in main
    rm.run_client(cfg, comm, model, loss_fn, args.num_clients, train_datasets)
  File "/gpfs/fs1/home/mryu/APPFL/src/appfl/run_mpi.py", line 283, in run_client
    ls = client.update()                            
  File "/gpfs/fs1/home/mryu/APPFL/src/appfl/algorithm/iiadmm.py", line 158, in update
    output = self.model(data)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/gpfs/fs1/home/mryu/APPFL/examples/models/resnet.py", line 65, in forward
    x = self.layer1(x)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/container.py", line 217, in forward
    input = module(input)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torchvision/models/resnet.py", line 97, in forward
    out = self.bn2(out)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/batchnorm.py", line 171, in forward
    return F.batch_norm(
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/functional.py", line 2450, in batch_norm
    return torch.batch_norm(
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 2.00 MiB (GPU 0; 39.59 GiB total capacity; 11.14 GiB already allocated; 4.19 MiB free; 11.27 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
[gpu2:3606526:0:3606526] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x14601a006ff8)
==== backtrace (tid:3606526) ====
 0  /soft/ucx/1.13.1/gcc-9.4.0/lib/libucs.so.0(ucs_handle_error+0x2dc) [0x14636aa48f6c]
 1  /soft/ucx/1.13.1/gcc-9.4.0/lib/libucs.so.0(+0x3414f) [0x14636aa4914f]
 2  /soft/ucx/1.13.1/gcc-9.4.0/lib/libucs.so.0(+0x34484) [0x14636aa49484]
 3  /lib/x86_64-linux-gnu/libc.so.6(+0x18e885) [0x14644a66d885]
 4  /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/libopen-pal.so.40(opal_convertor_pack+0xc1) [0x14636bb8da41]
 5  /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/openmpi/mca_btl_smcuda.so(mca_btl_smcuda_prepare_src+0xe3) [0x14636b0061c3]
 6  /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/openmpi/mca_pml_ob1.so(mca_pml_ob1_send_request_schedule_once+0x1dd) [0x14636a95748d]
 7  /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/openmpi/mca_pml_ob1.so(mca_pml_ob1_recv_frag_callback_ack+0x1b9) [0x14636a94f7d9]
 8  /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/openmpi/mca_btl_smcuda.so(mca_btl_smcuda_component_progress+0x426) [0x14636b008d46]
 9  /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/libopen-pal.so.40(opal_progress+0x34) [0x14636bb7c1e4]
10  /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/libmpi.so.40(ompi_mpi_finalize+0x1ad) [0x14636bd6bc0d]
11  /home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/mpi4py/MPI.cpython-38-x86_64-linux-gnu.so(+0x531bf) [0x14636bebb1bf]
12  python() [0x453de2]
13  python(Py_RunMain+0x112) [0x5a1972]
14  python(Py_BytesMain+0x39) [0x579dd9]
15  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x14644a5060b3]
16  python() [0x579c8d]
=================================
[gpu2:3606526] *** Process received signal ***
[gpu2:3606526] Signal: Segmentation fault (11)
[gpu2:3606526] Signal code:  (-6)
[gpu2:3606526] Failing at address: 0x4f8d003707fe
[gpu2:3606526] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0)[0x14644a8423c0]
[gpu2:3606526] [ 1] /lib/x86_64-linux-gnu/libc.so.6(+0x18e885)[0x14644a66d885]
[gpu2:3606526] [ 2] /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/libopen-pal.so.40(opal_convertor_pack+0xc1)[0x14636bb8da41]
[gpu2:3606526] [ 3] /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/openmpi/mca_btl_smcuda.so(mca_btl_smcuda_prepare_src+0xe3)[0x14636b0061c3]
[gpu2:3606526] [ 4] /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/openmpi/mca_pml_ob1.so(mca_pml_ob1_send_request_schedule_once+0x1dd)[0x14636a95748d]
[gpu2:3606526] [ 5] /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/openmpi/mca_pml_ob1.so(mca_pml_ob1_recv_frag_callback_ack+0x1b9)[0x14636a94f7d9]
[gpu2:3606526] [ 6] /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/openmpi/mca_btl_smcuda.so(mca_btl_smcuda_component_progress+0x426)[0x14636b008d46]
[gpu2:3606526] [ 7] /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/libopen-pal.so.40(opal_progress+0x34)[0x14636bb7c1e4]
[gpu2:3606526] [ 8] /soft/openmpi/4.1.4/swing/gcc-9.4/cuda-11.4/ucx-1.13.1/lib/libmpi.so.40(ompi_mpi_finalize+0x1ad)[0x14636bd6bc0d]
[gpu2:3606526] [ 9] /home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/mpi4py/MPI.cpython-38-x86_64-linux-gnu.so(+0x531bf)[0x14636bebb1bf]
[gpu2:3606526] [10] python[0x453de2]
[gpu2:3606526] [11] python(Py_RunMain+0x112)[0x5a1972]
[gpu2:3606526] [12] python(Py_BytesMain+0x39)[0x579dd9]
[gpu2:3606526] [13] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)[0x14644a5060b3]
[gpu2:3606526] [14] python[0x579c8d]
[gpu2:3606526] *** End of error message ***
./femnist.py:81: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:245.)
  torch.FloatTensor(test_data_input), torch.tensor(test_data_label)
Traceback (most recent call last):
  File "./femnist.py", line 169, in <module>
    main()
  File "./femnist.py", line 162, in main
    rm.run_client(cfg, comm, model, loss_fn, args.num_clients, train_datasets)
  File "/gpfs/fs1/home/mryu/APPFL/src/appfl/run_mpi.py", line 283, in run_client
    ls = client.update()                            
  File "/gpfs/fs1/home/mryu/APPFL/src/appfl/algorithm/iiadmm.py", line 158, in update
    output = self.model(data)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/gpfs/fs1/home/mryu/APPFL/examples/models/resnet.py", line 68, in forward
    x = self.layer4(x)                
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/container.py", line 217, in forward
    input = module(input)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torchvision/models/resnet.py", line 93, in forward
    out = self.bn1(out)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/batchnorm.py", line 171, in forward
    return F.batch_norm(
  File "/home/mryu/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/functional.py", line 2450, in batch_norm
    return torch.batch_norm(
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 20.00 MiB (GPU 0; 39.59 GiB total capacity; 11.73 GiB already allocated; 8.19 MiB free; 11.76 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 1 with PID 3606526 on node gpu2 exited on signal 11 (Segmentation fault).
--------------------------------------------------------------------------

```

**Comment by yim0331** [comment]
The out-of-memory error occurs by communication via GPU memory. 

The solution that we used for the client_optimizer.py was moving the local_state data to the CPU memory by the following code:
https://github.com/APPFL/APPFL/blob/e4a7509aa9a5001b3e38e56ce00c56e21ee238fd/src/appfl/algorithm/client_optimizer.py#L107-L110

However, the iiadmm and iceadmm client class seems not to have this kind of moving code.

**Comment by minseok-ryu** [comment]
ICEADMM now works.
Command:
```
python ./femnist.py --model resnet18 --pretrained 1 --fed="ICEADMM()"
```
Log:
```
      Iter   Local(s)  Global(s)   Valid(s) PerIter(s) Elapsed(s)   TestLoss   TestAccu     PrimRes      DualRes       RhoMin       RhoMax
         1      33.59       9.88       4.63      48.10      48.10    11.7661       0.00  8.6218e-03   0.0000e+00   1.0000e+02   3.2000e+03
         2      32.46      11.64       4.62      48.72      96.82    11.7661       0.00  1.3714e-03   2.1921e+03   4.0000e+02   4.0960e+05
Device=cuda
#Processors=1
#Clients=197
Server=ICEADMMServer
Clients=ICEADMMClient
Comm_Rounds=2
Local_Rounds=1
DP_Eps=False
Clipping=False
Elapsed_time=96.82
BestAccuracy=0.0
```

File: examples/femnist.py [hunk 1]
```diff
@@ -27,6 +28,8 @@
 parser.add_argument("--num_pixel", type=int, default=28)
 parser.add_argument("--model", type=str, default="CNN")
 
+## algorithm
+parser.add_argument("--fed", type=str, default="Federated()")  ## Federated(), ICEADMM(), IIADMM()
```
Comment (kibaekkim) [inline examples/femnist.py:32]: Should we come up with a better name for the argument? `--federation_type`? Can we drop `()` from the argument values?

**Review by kibaekkim — COMMENTED** [review]
There are strange white spaces in the changes. Can you remove them? @yim0331 Did we/you use any formatter? Can you recommend one?

**Comment by yim0331** [comment]
I don't use specific formatter. However, I [TrailingSpaces](https://github.com/SublimeText/TrailingSpaces) to clean up the white spaces with Sublime Text.

**Review by yim0331 — APPROVED** [review]

File: examples/femnist.py (hunk 1 above)
Comment (minseok-ryu) [reply examples/femnist.py]: Reflected.

**Review by kibaekkim — APPROVED** [review]

**Pull request merged**